metadata/src/main/java/org/apache/kafka/controller/LogReplayTracker.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.controller;
+
+import java.util.Optional;
+
+import org.apache.kafka.common.protocol.ApiMessage;
+import org.apache.kafka.common.utils.LogContext;
+import org.apache.kafka.metadata.bootstrap.BootstrapMetadata;
+import org.apache.kafka.server.common.MetadataVersion;
+import org.slf4j.Logger;
+
+
+public class LogReplayTracker {
+    public static class Builder {
+        private LogContext logContext = null;
+
+        Builder setLogContext(LogContext logContext) {
+            this.logContext = logContext;
+            return this;
+        }
+
+        public LogReplayTracker build() {
+            if (logContext == null) logContext = new LogContext();
+            return new LogReplayTracker(logContext);
+        }
+    }
+
+    /**
+     * The slf4j logger.
+     */
+    private final Logger log;
+
+    /**
+     * True if we haven't replayed any records yet.
+     */
+    private boolean empty;
+
+    /**
+     * True if we haven't replayed any metadata.version records yet.
+     */
+    private boolean versionless;
```

Comment:
nit: the name confused me a little bit. How about `hasReadMetadataVersion` or something like that? Or perhaps even simpler, we could record the `MetadataVersion` itself.

Reply:
I usually don't push back on name suggestions but I really liked "versionless" ... it's a log without a version. Seems apt?

Same for "empty" -- it just feels right.

Of course these things are mutable, will change over time, but as a description of the current state it seems correct?